course_flow/tasks.py

```diff
@@ -0,0 +1,314 @@
+        ).distinct()
+        postreqs = Node.objects.filter(
+            incoming_links__source_node__in=nodes
+        ).distinct()
+        if len(prereqs) > 0:
+            df = df.append(
+                {
+                    "a": _("Prerequisites"),
+                    "b": [get_displayed_title(req) for req in prereqs].join(
+                        ", "
+                    ),
+                },
+                ignore_index=True,
+            )
+        if len(postreqs) > 0:
+            df = df.append(
+                {
+                    "a": _("Required For"),
+                    "b": [get_displayed_title(req) for req in postreqs].join(
+                        ", "
+                    ),
+                },
+                ignore_index=True,
+            )
+    df = df.append(
+        {
+            "a": _("Course Outcome"),
+            "b": _("Sub-Outcomes"),
+            "c": _("Competencies"),
+            "d": _("Topics & Content"),
+            "e": _("Lessons/Activities"),
+            "f": _("Assessments"),
+        },
+        ignore_index=True,
+    )
+    for outcome in workflow.outcomes.all():
+        df = df.append(
+            get_framework_line_for_outcome(outcome), ignore_index=True
+        )
+    return df
+
+
+def get_workflow_outcomes_table(workflow):
+    outcomes = get_all_outcomes_ordered(workflow)
+    data = OutcomeExportSerializer(outcomes, many=True).data
+    df = pd.DataFrame(
+        data, columns=["code", "title", "description", "id", "depth"]
+    )
+    pd.set_option("display.max_colwidth", None)
+    return df
+
+
+@shared_task
+def async_get_outcomes_excel(email, pk, object_type):
+    model_object = get_model_from_str(object_type).objects.get(pk=pk)
+    with BytesIO() as b:
+        # Use the StringIO object as the filehandle.
+        writer = pd.ExcelWriter(b, engine="openpyxl")
+        if object_type == "workflow":
+            workflows = [model_object]
+        elif object_type == "project":
+            workflows = list(model_object.workflows.all())
+        for workflow in workflows:
+            df = get_workflow_outcomes_table(workflow)
+            df.to_excel(
+                writer,
+                sheet_name=workflow.title + "_" + str(workflow.pk),
+                index=False,
+            )
+            writer.save()
+        # Set up the Http response.
+        filename = (
+            object_type
+            + "_"
+            + str(pk)
+            + "_"
+            + timezone.now().strftime(dateTimeFormatNoSpace())
+            + ".xlsx"
+        )
+        email = EmailMessage(
+            "Your Outcomes Export",
+            "Hi there! Here are the results of your recent outcomes export.",
+            "noreply@courseflow.org",
+            [email],
+        )
+        email.attach(
+            filename,
+            b.getvalue(),
+            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+        )
+        email.send()

```

Comment:
This should probably be wrapped in a `try` block and any SMTP exception handled.